### PR TITLE
virt-handler shared informers

### PR DESCRIPTION
**Pull request by davidvossel**
migrate virt-handler to use the same informer patterns as virt-controller.

### CODE REVIEW

**Comment by davidvossel** [comment]
retest this please

**Comment by davidvossel** [comment]
Hoping to avoid code rot here. Anyone feel comfortable giving this PR an ack?

**Comment by davidvossel** [comment]
actually, lets get the "non-default namespace" patch in before this. I'll have to merge with those changes.

**Comment by davidvossel** [comment]
@rmohr I updated this to work with non-default namespaces. should be good to go now.

File: pkg/informers/virtinformers.go [hunk 1]
```diff
@@ -108,6 +114,23 @@ func (f *kubeInformerFactory) getInformer(key string, newFunc newSharedInformer)
...
+
+func (f *kubeInformerFactory) VmOnHost(host string) cache.SharedIndexInformer {
+	name := fmt.Sprintf("vmOnHostInformer_%s", host)
+	return f.getInformer(name, func() cache.SharedIndexInformer {
+		labelSelector, err := labels.Parse(fmt.Sprintf(kubev1.NodeNameLabel+" in (%s)", host))
+		if err != nil {
+			panic(err)
+		}
+
+		lw := kubecli.NewListWatchFromClient(f.restClient, "vms", kubeapi.NamespaceDefault, fields.Everything(), labelSelector)
```
Comment (vladikr) [inline pkg/informers/virtinformers.go]: Just wondering if this shouldn't be a api.NamespaceAll?

File: pkg/virt-handler/virtwrap/cache/cache.go [hunk 2]
```diff
@@ -70,7 +70,7 @@ var CrashedReasonTranslationMap = map[libvirt.DomainCrashedReason]api.StateChang
 }
 
 // NewListWatchFromClient creates a new ListWatch from the specified client, resource, namespace and field selector.
-func newListWatchFromClient(c virtwrap.Connection, events ...int) *cache.ListWatch {
+func NewListWatchFromClient(c virtwrap.Connection, events ...int) *cache.ListWatch {
```
Comment (rmohr) [inline pkg/virt-handler/virtwrap/cache/cache.go]: Not that we were taking care about exports so far, but are you deliberately exporting that function?

File: cmd/virt-handler/virt-handler.go [hunk 3]
```diff
@@ -102,42 +102,36 @@ func main() {
...
 	}
 
-	l, err := labels.Parse(fmt.Sprintf(v1.NodeNameLabel+" in (%s)", *host))
-	if err != nil {
-		panic(err)
-	}
+	// Create all the informers used by virt-handler here
+	informerFactory := kubeinformers.NewKubeInformerFactory(restClient, clientSet)
+	vmOnHostInformer := informerFactory.VmOnHost(*host)
+	domainSharedInformer := informerFactory.CustomInformer("domainInformer", func() cache.SharedIndexInformer {
```
Comment (rmohr) [inline cmd/virt-handler/virt-handler.go]: Ok, when thinking about this, I don't really see a reason, to move this to a cluster wide factory. It is purely virt-handler and libvirt related code.

Comment (davidvossel) [inline cmd/virt-handler/virt-handler.go]: The domain informer is specific to virt-handler, that's why I created the "CustomInformer" function.  It lets the InformerFactory manage shared informers that should not be exposed cluster wide.

File: pkg/virt-handler/virtwrap/cache/cache.go (hunk 2 above)
Comment (davidvossel) [inline pkg/virt-handler/virtwrap/cache/cache.go]: yep, it's used to create the shared informer somewhere else.

**Comment by davidvossel** [comment]
> I am not really sure, if this PR is a good thing. We are moving virt-handler and libvirt related code to the same place, where cluster related informer code is constructed, which can be shared across the whole cluster. That kind of obfuscates, that the domain informer, is something node and virt-handler specific ...

The CustomInformer function keeps the domain informer from being something that is shared cluster wide.  This keeps something like 'virt-controller' from having to pull in any libvirt related client libraries.

**Comment by davidvossel** [comment]
closing for now. I don't feel super strongly about this patch. It isn't that valuable of a change.

**Pull request closed without merging**